### ISSUE TITLE
Accessibility (a11y) improvements for visual disabilities and color contrast

### ISSUE THREAD

**Issue by endigo9740** [skeletonlabs/skeleton#151]
The current release of Skeleton provides support for tab focus, keyboard shortcuts, semantic elements, and of course implements changes to support the ARAI spec per each component.

That said, we still have several items to address in regards to a11y support. The the next major update should include further refinement and support for visual impairments. This includes but is not limited to:

* Support for color blindness
* Support for low visibility or visual impaired users
* Improve color contrast support - note this may involve introducing new and dedicated themes

We should consider making suggestions for tools dedicated to testing. These can include:

* [axe DevTools](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US)
* [Web Disability Simulator](https://chrome.google.com/webstore/detail/web-disability-simulator/olioanlbgbpmdlgjnnampnnlohigkjla?hl=en)

Finally, this will be a bit less of a concrete fix, as color values are completely in control of the users generating the theme. We may want to make considerations for improving a11y support for our theme, or creating separate/dedicated themes specifically built with a11y in mind. Likewise we should update documentation or provide recommendations for a11y (ex: "be mindful of color contrast radio when overriding color values", etc)

_Update_

Great find for a11y color generation and testing here:
https://stephaniewalter.design/blog/color-accessibility-tools-resources-to-design-inclusive-products/

Dark Theme recommendations:
https://twitter.com/vitalyf/status/1564607043406815233

Leonardo:
https://leonardocolor.io/

**Comment by endigo9740** [comment]
Here's some notes after some reseach and testing for visual a11y. In short, it's not really about what color values are used, so much as how high the contrast ratio is. For our themes, the biggest offers is typically text on top of a primary or accent color.

Take this example, generated with https://huewind.jordantwells.com/

![Screen Shot 2022-08-25 at 3 11 28 PM](https://user-images.githubusercontent.com/1509726/186759513-950ec15d-dd58-4108-9d85-e32d0ddac4fe.png)

Note there's not a particular swatch level (ex: 500) that the overlapping text flips from white to black or vice versa. Rather this is handled per shade and varies per each column.

With this in mind, this will be hard handle this programmatically without an in-depth review of each and every single theme. It would be an insane effort to test every possible theme and component combination. Instead I propose we approach these challenges from a different angle. See my suggestions below.

## 1.
We improve our global and component default styles as much as possible. Aiming to better accommodate for contrast. Take our global paragraph style for example, the contrast is way too low, but can be easily fixed:

```css
main p { @apply text-base text-surface-500 dark:text-surface-400; }
```

Should be:

```css
main p { @apply text-base text-black/80 dark:text-white/80; }
```

Results before vs after:

![Screen Shot 2022-08-25 at 3 28 39 PM](https://user-images.githubusercontent.com/1509726/186762743-978b5a32-9a51-4438-bc4d-305a2f8ec0af.png)

![Screen Shot 2022-08-25 at 3 28 47 PM](https://user-images.githubusercontent.com/1509726/186762746-f6566f17-095e-4129-80df-d412d3f255b7.png)

This provides a much better contrast ratio out of the box for both light and dark modes. Though obviously users can choose to customize further if they wish. Perhaps even going for full 100% opacity.

## 2.
We may need to recommend that if a11y is a focus, you should pin your design to use only light or dark, not toggle between both. The variance between the two modes means there's too many variables to keep up with. Plus focusing one mode makes it easier to define and override the default values. For example, if we only used dark mode for the Skeleton theme, we could use the following to improve button contrast:

```css
.button { @apply !bg-primary-400 !text-black; }
```
Results before vs after:

![Screen Shot 2022-08-25 at 3 34 33 PM](https://user-images.githubusercontent.com/1509726/186763660-9d24dd65-99a0-416b-a728-735cb85cc5f3.png)

However, we should look to review each component and ensure it's possible to target styles in a global manner. The style above would affect ALL buttons, whether filled or outlined, which is not very visually appealing.

![Screen Shot 2022-08-25 at 3 49 13 PM](https://user-images.githubusercontent.com/1509726/186766040-4a8b6fb7-8364-4312-9898-c8d3b6a88316.png)

## 3.

Finally, we should provide at least one theme and set of stylesheets purpose-built for a11y. This could be our reference for best practices. Per this, most a11y-focused themes I've seen to to use high contrasting black/white theme with subtle color highlights for interactive elements (ex: buttons). I'd expect we should follow this example below:
https://labinator.com/wordpress-marketplace/lat-demo/

**Comment by endigo9740** [comment]
I'd advise we handle this after the built-in themes and modular stylesheet update is merged into dev branch: https://github.com/Brain-Bones/skeleton/pull/137

This will make this process MUCH easier!

**Comment by endigo9740** [comment]
Note that the new proposed Button Refactor opens the door to handling button visual a11y by allowing for easier global overrides of styling:

https://github.com/Brain-Bones/skeleton/issues/216

**Comment by endigo9740** [comment]
Note that we've also update the default typography to improve color contrast rations. So a bit more progress here!

**Comment by ryceg** [comment]
It's worth noting that the `@prefers-contrast` media query (`:contrast-more` in Tailwind) has support for both `less` and `more`, and might be a good option for making things more a11y friendly.

**Comment by endigo9740** [comment]
@ryceg Really great info to know! Thanks!

I am still relatively new to the a11y side of things, but trying to learn and improve. Any other tips are absolutely welcome, visual or otherwise.

**Comment by ryceg** [comment]
We can use the formula listed on [WCAG](https://www.w3.org/TR/WCAG20-TECHS/G17.html) and test whether colours meet accessibility guidelines, and generate complementary "high contrast" colours. We'd also be able to provide users with this information while they're generating themes, which would be _very_ helpful. I won't have time to dig into this just yet, so don't assign it to me in case someone else wants to have a crack at it, but I figured I'd at least put that idea out there so it's visible.

**Comment by endigo9740** [comment]
https://twitter.com/vitalyf/status/1580017289868771328

**Comment by endigo9740** [comment]
With the following update going live tomorrow, I'm feeling confident in our visual a11y:
https://github.com/skeletonlabs/skeleton/issues/553

This doesn't meant things are perfect or 100% accessible, but rather that we're meeting the standards and quality expected when this ticket was created way back in August. I do expect we'll see additional updates and improvements as part of these tickets:

https://github.com/skeletonlabs/skeleton/issues/374

Likewise another dedicated a11y audit will be recommend prior to the v1.0 launch.

But for now, this ticket has served it's purpose.
